--- a/elfeed-search.el	1970-01-01 03:00:01.000000000 +0300
+++ b/elfeed-search.el	2020-08-25 22:02:04.114008818 +0300
@@ -233,7 +233,6 @@
   (set (make-local-variable 'bookmark-make-record-function)
        #'elfeed-search-bookmark-make-record)
   (buffer-disable-undo)
-  (hl-line-mode)
   (make-local-variable 'elfeed-search-entries)
   (make-local-variable 'elfeed-search-filter)
   (add-hook 'elfeed-update-hooks #'elfeed-search-update)
@@ -766,14 +765,14 @@
   (interactive "P")
   (let ((entries (elfeed-search-selected)))
     (cl-loop for entry in entries
-             do (elfeed-untag entry 'unread)
+             do (progn (elfeed-untag entry 'unread)
+                       (mapc #'elfeed-search-update-entry entries)
+                       (unless (or elfeed-search-remain-on-entry (use-region-p))
+                         (forward-line)))
              when (elfeed-entry-link entry)
              do (if use-generic-p
                     (browse-url-generic it)
-                  (browse-url it)))
-    (mapc #'elfeed-search-update-entry entries)
-    (unless (or elfeed-search-remain-on-entry (use-region-p))
-      (forward-line))))
+                  (browse-url it)))))
 
 (defun elfeed-search-yank ()
   "Copy the selected feed items to clipboard and kill-ring."
