--- a/info-colors.el	1970-01-01 03:00:01.000000000 +0300
+++ b/info-colors.el	2018-02-07 20:25:59.187852988 +0300
@@ -120,7 +120,7 @@
   (goto-char (point-min))
   (while (re-search-forward
           "^ --? \
-\\(Command\\|Constant\\|Function\\|Macro\\|Special Form\\|Syntax class\\|User Option\\|Variable\\):\
+\\(Command\\|Constant\\|C Function\\|Function\\|Scheme Procedure\\|Macro\\|Special Form\\|Syntax\\|Syntax class\\|User Option\\|Variable\\):\
  *\\(\\S-+\\)\
 \\(\\( .*\\)?\\([\n] \\{8\\}.*\\)*\\)"
           nil t)
